agent/src/util/discovery_operator.rs

```diff
@@ -0,0 +1,1504 @@
+    /// if it have been offline for longer than the grace period.
+    /// If `DiscoveryHandlerConnectivityStatus::HasClient`, this should not happen, Error is returned.
+    pub async fn mark_offline_or_deregister_discovery_handler(
+        &self,
+        endpoint: &DiscoveryHandlerEndpoint,
+    ) -> Result<bool, anyhow::Error> {
+        trace!("mark_offline_or_deregister_discovery_handler - discovery handler at endpoint {:?} and protocol {} is offline", endpoint, self.config.spec.protocol.name);
+        let mut deregistered = false;
+        let mut registered_dh_map = self.discovery_handler_map.lock().unwrap();
+        let protocol_map = registered_dh_map
+            .get_mut(&self.config.spec.protocol.name)
+            .unwrap();
+        let dh_details = protocol_map.get_mut(endpoint).unwrap();
+        match dh_details.connectivity_status {
+            DiscoveryHandlerConnectivityStatus::Offline(instant) => {
+                if instant.elapsed().as_secs() > DISCOVERY_HANDLER_OFFLINE_GRACE_PERIOD_SECS {
+                    trace!("mark_offline_or_deregister_discovery_handler - de-registering discovery handler for protocol {} at endpoint {:?} since been offline for longer than 5 minutes", self.config.spec.protocol.name, endpoint);
+                    // Remove discovery handler from map if timed out
+                    protocol_map.remove(endpoint).unwrap();
+                    deregistered = true;
+                }
+            }
+            DiscoveryHandlerConnectivityStatus::Online => {
+                dh_details.connectivity_status =
+                    DiscoveryHandlerConnectivityStatus::Offline(Instant::now());
+            }
+            DiscoveryHandlerConnectivityStatus::HasClient => {
+                dh_details.connectivity_status =
+                    DiscoveryHandlerConnectivityStatus::Offline(Instant::now());
+            }
+        }
+        Ok(deregistered)
+    }
+
+    /// Checks if any of this DiscoveryOperator's Configuration's Instances have been offline for too long.
+    /// If a non-local device has not come back online before `SHARED_INSTANCE_OFFLINE_GRACE_PERIOD_SECS`,
+    /// the associated Device Plugin and Instance are terminated and deleted, respectively.
+    #[allow(dead_code)]
+    pub async fn delete_offline_instances(
+        &self,
+        kube_interface: Arc<Box<dyn k8s::KubeInterface>>,
+    ) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+        trace!(
+            "delete_offline_instances - entered for configuration {}",
+            self.config.metadata.name
+        );
+        let kube_interface_clone = kube_interface.clone();
+        let instance_map = self.instance_map.lock().await.clone();
+        for (instance, instance_info) in instance_map.clone() {
+            match instance_info.connectivity_status {
+                InstanceConnectivityStatus::Online => {}
```

Comment:
should HasClient be treated as an error here?

Reply:
of should it be:

```
InstanceConnectivityStatus::Online | InstanceConnectivityStatus::HasClient => {} // should we log this?
```

Reply:
This is called periodically to delete **Instances** that have been offline for too long. HasClient is associated with **discovery handlers**. I dont see a reason to log the online scenario. An if Let might be better than a match here